scripts/firmware.py

```diff
@@ -0,0 +1,2097 @@
+        """Execute the action associated with a valid menu choice."""
+        menu_item = self.menu_items[choice]
+        if isinstance(menu_item, self.MenuItem):
+            menu_item.action()  # Safe to call because type is now narrowed
+        else:
+            raise TypeError(f"Expected MenuItem, got {type(menu_item).__name__}")
+
+    def invalid(self) -> None:
+        """Display a message for an invalid choice."""
+        Utils.header()
+        print(Utils.colored_text("Invalid choice. Please try again.", Color.RED))
+
+
+class Validator:
+    """A utility class for common validation checks."""
+
+    def __init__(self, firmware: "Firmware"):
+        self.firmware: Firmware = (
+            firmware  # Reference to the firmware object for navigation
+        )
+
+    def validate_device(self, device: str, type: str) -> bool:
+        if type == "CAN":
+            device_regex = r"^[a-f0-9]{12}$"
+        elif type == "USB":
+            device_regex = r".*Cartographer.*"
+        else:
+            device_regex = r"^[a-f0-9]{4}:[a-f0-9]{4}$"
+        return bool(re.match(device_regex, device))
+
+    def check_selected_firmware(self):
+        if not self.firmware.selected_firmware:
+            self._error_and_return("You have not selected a firmware file.")
+
+    def check_selected_device(self):
+        if not self.firmware.selected_device:
+            self._error_and_return("You have not selected a device to flash.")
+
+    def check_temp_directory(self):
+        if self.firmware.dir_path is None:
+            self._error_and_return("Error getting temporary directory path.")
+
+    def _error_and_return(self, message: str):
+        Utils.error_msg(message)
+        _ = input(
+            Utils.colored_text(
+                "\nPress Enter to return to the main menu...", Color.YELLOW
+            )
+        )
+        self.firmware.main_menu()
+
+
+class Firmware:
+    can: "Can"
+    usb: "Usb"
+    dfu: "Dfu"
+
+    def __init__(
+        self,
+        branch: str = "master",
+        debug: bool = False,
+        ftype: bool = False,
+        high_temp: bool = False,
+        flash: Optional[str] = None,
+        kseries: bool = False,
+        all: bool = False,
+        device: Optional[str] = None,
+    ):
+        self.selected_device: Optional[str] = None
+        self.selected_firmware: Optional[str] = None
+        self.dir_path: Optional[str] = None
+        self.debug: bool = debug
+        self.branch: str = branch
+        self.ftype: bool = ftype
+
+        self.high_temp: bool = high_temp
+        self.flash: Optional[str] = flash
+        self.kseries: bool = kseries
+        self.all: bool = all
+        self.device: Optional[str] = device
+        self.can = Can(self, debug=self.debug, ftype=self.ftype)
+        self.usb = Usb(self, debug=self.debug, ftype=self.ftype)
+        self.dfu = Dfu(
+            self, debug=self.debug, ftype=self.ftype
+        )  # Pass Firmware instance to CAN

```

Comment:
This is string type on the class instance, now it's some other thing.
What is this?